fms/utils/generation.py

```diff
@@ -1,7 +1,106 @@
-from typing import Any, Callable, List, MutableMapping, Optional, Union
+import inspect
+from typing import Any, Callable, Dict, List, MutableMapping, Optional, Tuple, Union
 
 import torch
+import torch.nn as nn
 import torch.nn.functional as F
+from torch._dynamo import OptimizedModule
+
+
+def __prepare_list_input(
+    input_ids_list: List[torch.Tensor], model: Union[Callable, torch.nn.Module]
+) -> Tuple[torch.Tensor, MutableMapping[str, Any]]:
+    """
+    Convert the list of Tensors to a rectangular tensor. Return extra kwargs for the position_ids and mask, since this
+    will be required to properly handle the rectangular tensor for certain models.
+    """
+    min_len = min([seq.size(0) for seq in input_ids_list])
+    max_len = max([seq.size(0) for seq in input_ids_list])
+
+    if isinstance(model, OptimizedModule):
+        forward_function = model._orig_mod.forward
+    elif isinstance(model, nn.Module):
+        forward_function = model.forward
+    else:
+        forward_function = model
+
+    params = inspect.signature(forward_function).parameters.keys()
+    extra_kwargs = {}
+    needs_mask = "mask" in params and min_len != max_len
+    needs_position_ids = "position_ids" in params and min_len != max_len
+
+    padded_input_ids_list = []
+    mask_list = []
+    position_ids_list = []
+    for input_ids_i in input_ids_list:
+        seq_len = input_ids_i.size(0)
+        pads = torch.zeros(
+            max_len - seq_len, dtype=torch.long, device=input_ids_i.device
+        )
+        non_pads = torch.ones(seq_len, dtype=torch.bool, device=input_ids_i.device)
+
+        # Setting this to 0, however if 0 is the eos, we will end up truncating the output if using truncate_after_eos
+        # once this workflow works for nested tensor, this can probably be removed
+        padded_input_ids_list.append(torch.cat((pads, input_ids_i)))
+
+        # computing this as it's lightweight but could potentially be skipped
+        mask_list.append(torch.cat((pads.bool(), non_pads)))
+        position_ids_list.append(
+            torch.cat(
+                (
+                    pads,
+                    torch.arange(
+                        0, seq_len, dtype=torch.long, device=input_ids_i.device
+                    ),
```

Comment:
this code formatting could probably be made easier to read as something like 
```
new_pads = ...
new_pos_ids = ...
position_ids.append(new_pos_ids)
```
though I"m a bit confused by what "pads" means here and maybe a better name could make that clearer too

Reply:
Fixed in latest commit